### PR TITLE
Fix(TDI-41289):Add reuse tS3Connection.

### PR DESCRIPTION
**Pull request by vyu-talend**
**What is the current behavior?** (You can also link to an open issue here)
https://jira.talendforge.org/browse/TDI-41289

**What is the new behavior?**

**Please check if the PR fulfills these requirements**

- [ ] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [ ] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: main/plugins/org.talend.designer.components.localprovider/components/tRedshiftBulkExec/tRedshiftBulkExec_begin.javajet [hunk 1]
```diff
@@ -221,13 +226,30 @@ skeleton="../templates/db_output_bulk.skeleton"
...
+%>
+			final String decryptedPwd_<%=cid%> = <%= ElementParameterParser.getValue(node, passwordFieldKey)%>; 
+<%
+		}
+	} else {
+		access_key = ElementParameterParser.getValue(s3ConnectionNode, "__ACCESS_KEY__");
+    	passwordFieldKey = "__SECRET_KEY__";
+		if (ElementParameterParser.canEncrypt(s3ConnectionNode, passwordFieldKey)) {
+%>
+			final String decryptedPwd_<%=cid%> = routines.system.PasswordEncryptUtil.decryptPassword(<%=ElementParameterParser.getEncryptedValue(s3ConnectionNode, passwordFieldKey)%>);
```
Comment (pyzhou-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tRedshiftBulkExec/tRedshiftBulkExec_begin.javajet]: better to reduce such kind of replication as far as possible. Like make the judgment at the very beginning and assign the "currect" value for access_key and passwordFieldKey.

File: main/plugins/org.talend.designer.components.localprovider/components/tRedshiftOutputBulkExec/tRedshiftOutputBulkExec_java.xml [hunk 2]
```diff
@@ -124,12 +124,20 @@
...
+		<PARAMETER NAME="USE_EXISTING_CONNECTION_S3" FIELD="CHECK"
+	        NUM_ROW="46"  GROUP="S3_CONFIG">
+	        <DEFAULT>false</DEFAULT>
+	    </PARAMETER>
+
+		<PARAMETER NAME="CONNECTION_S3" FIELD="COMPONENT_LIST" REQUIRED="true" FILTER="tS3Connection"
+        	NUM_ROW="46" SHOW_IF="USE_EXISTING_CONNECTION_S3 == 'true'" GROUP="S3_CONFIG"/>
+
         <PARAMETER
           NAME="ACCESS_KEY"
```
Comment (jzhao-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tRedshiftOutputBulkExec/tRedshiftOutputBulkExec_java.xml]: you need to recheck the what you reused parameters from **ts3connection** for **tredshfitoutputbulkexec**
add update the comment in https://jira.talendforge.org/browse/TDI-41289 for doct team
you also need to check other components, you need list all the parameters which is reused.

File: main/plugins/org.talend.designer.components.localprovider/components/tRedshiftBulkExec/tRedshiftBulkExec_begin.javajet [hunk 3]
```diff
@@ -221,13 +226,30 @@ skeleton="../templates/db_output_bulk.skeleton"
...
+		if (ElementParameterParser.canEncrypt(node, passwordFieldKey)) {
+%> 
+			final String decryptedPwd_<%=cid%> = routines.system.PasswordEncryptUtil.decryptPassword(<%=ElementParameterParser.getEncryptedValue(node, passwordFieldKey)%>);
+<%
+		} else {
+%>
+			final String decryptedPwd_<%=cid%> = <%= ElementParameterParser.getValue(node, passwordFieldKey)%>; 
+<%
+		}
+	} else {
```
Comment (pyzhou-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tRedshiftBulkExec/tRedshiftBulkExec_begin.javajet]: Better to reduce such kind of replication as far as possible. Why not just judge and assign the correct value to variables accessKey and password at the very beginning?

File: main/plugins/org.talend.designer.components.localprovider/components/tRedshiftBulkExec/tRedshiftBulkExec_begin.javajet [hunk 4]
```diff
@@ -298,20 +320,38 @@ skeleton="../templates/db_output_bulk.skeleton"
 <%
 	}
 	boolean encrypt = "true".equals(ElementParameterParser.getValue(node, "__ENCRYPT__"));
-	encrypt = encrypt && s3Mode;
-	if(encrypt) {
-		String encryptionKeyPropertyName = "__ENCRYPTED_KEY__";
+	boolean encryptS3Conn = "true".equals(ElementParameterParser.getValue(s3ConnectionNode, "__ENCRYPT__"));
```
Comment (pyzhou-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tRedshiftBulkExec/tRedshiftBulkExec_begin.javajet]: Similar to above, what the difference between encrypt and encryptS3conn?

File: main/plugins/org.talend.designer.components.localprovider/components/tRedshiftBulkExec/tRedshiftBulkExec_begin.javajet [hunk 5]
```diff
@@ -221,13 +226,30 @@ skeleton="../templates/db_output_bulk.skeleton"
...
 %>
-	<%if (ElementParameterParser.canEncrypt(node, passwordFieldKey)) {%> 
-	final String decryptedPwd_<%=cid%> = routines.system.PasswordEncryptUtil.decryptPassword(<%=ElementParameterParser.getEncryptedValue(node, passwordFieldKey)%>);
-	<%} else {%>
-	final String decryptedPwd_<%=cid%> = <%= ElementParameterParser.getValue(node, passwordFieldKey)%>; 
-	<%}%>
-	
 <%
+	if(!useExistingS3Connection){
+		if (ElementParameterParser.canEncrypt(node, passwordFieldKey)) {
```
Comment (pyzhou-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tRedshiftBulkExec/tRedshiftBulkExec_begin.javajet]: I know it used to be like this. But better to introduce to a local variable for readability.

File: main/plugins/org.talend.designer.components.localprovider/components/tRedshiftOutputBulkExec/tRedshiftOutputBulkExec_java.xml (hunk 2 above)
Comment (vyu-talend) [reply main/plugins/org.talend.designer.components.localprovider/components/tRedshiftOutputBulkExec/tRedshiftOutputBulkExec_java.xml]: I will recheck it and introduce the current behavior to DOCT.

File: main/plugins/org.talend.designer.components.localprovider/components/tRedshiftBulkExec/tRedshiftBulkExec_begin.javajet (hunk 5 above)
Comment (vyu-talend) [reply main/plugins/org.talend.designer.components.localprovider/components/tRedshiftBulkExec/tRedshiftBulkExec_begin.javajet]: Thank you, I will review it.

File: main/plugins/org.talend.designer.components.localprovider/components/tRedshiftOutputBulk/tRedshiftOutputBulk_end.javajet [hunk 6]
```diff
@@ -36,15 +36,23 @@
...
-			%>	
+			%>
 				log.info("<%=cid%> - Upload the object successfully.");
 			<%
 			}
 			%>
+		} catch(java.lang.Exception e_<%=cid%>) {
+			throw new java.lang.Exception(e_<%=cid%>.getMessage());
+			<%
+			if(!("true").equals(useExistingConn)) {
```
Comment (pyzhou-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tRedshiftOutputBulk/tRedshiftOutputBulk_end.javajet]: where does this variable come from? I didn't find it.
And the catch block is meaningless here. You can just control the generation of 
`if(conn_<%=cid%> != null){
conn_<%=cid%>.shutdown();
}`

File: main/plugins/org.talend.designer.components.localprovider/components/tRedshiftOutputBulk/tRedshiftOutputBulk_end.javajet [hunk 7]
```diff
@@ -36,15 +36,23 @@
...
 			<%
 			if(isLog4jEnabled){
-			%>	
+			%>
 				log.info("<%=cid%> - Upload the object successfully.");
 			<%
 			}
 			%>
+		} catch(java.lang.Exception e_<%=cid%>) {
+			throw new java.lang.Exception(e_<%=cid%>.getMessage());
```
Comment (pyzhou-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tRedshiftOutputBulk/tRedshiftOutputBulk_end.javajet]: Although you should eliminate this, throw e_<%=cid%> is better. Not need to instance a new Exception.

File: main/plugins/org.talend.designer.components.localprovider/components/tRedshiftUnload/tRedshiftUnload_begin.javajet [hunk 8]
```diff
@@ -163,25 +179,43 @@ imports="
 <%
 	}
 	boolean encrypt = "true".equals(ElementParameterParser.getValue(node, "__ENCRYPT__"));
-	if(encrypt) {
-		String encryptionKeyPropertyName = "__ENCRYPTED_KEY__";
+	boolean encryptS3Conn = "true".equals(ElementParameterParser.getValue(s3ConnectionNode, "__ENCRYPT__"));
```
Comment (pyzhou-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tRedshiftUnload/tRedshiftUnload_begin.javajet]: Should modify this same as tRedshiftBulkExec.

**Pull request merged**